### ISSUE TITLE
Multiple requests to retrieveLocale

### ISSUE THREAD

**Issue by NateWr** [Juris-M/citeproc-js#41]
I'm working on integrating `citeproc-js` with OJS 3, an open-source journal publishing platform. We want to use it to display a citation for a published article, and allow users to generate a citation in one of a handful of formats the journal manager might enable.

One of my goals is to minimise the performance impact of loading the citation on an end-user-facing page. To do this, I load the locale and default style with the initial payload, and bootstrap the first render of a citation from XML files loaded into a global var. So my `citeprocSys` retrieves locale and item data from an existing object, rather than fetching it on-demand:

```js
var citeprocSys = {
	retrieveLocale: function(lang) {
		if (typeof pkpCSL.locales[lang] !== 'undefined' ) {
			return pkpCSL.locales[lang];
		}
		return false;
	},
	retrieveItem: function(id) {
		return pkpCSL.items[id];
	}
};
```

This works because only a single language will be available per page request, so we don't need to load in all the XML every time.

One issue I've found with this is that `citeproc.js` fetches an `en-US` locale every time, regardless of what language is requested, even if the requested language file exists.

https://github.com/Juris-M/citeproc-js/blob/master/src/util_locale.js#L40-L49

I wonder if it would be possible to fetch the fallback locales only when the requested locale is not available. `retrieveLocale` could then return false when a locale is not available. This would prevent `retrieveLocale` from getting called multiple times when unnecessary.

The code for that could look like this:

```js
    localexmlraw = this.sys.retrieveLocale(langspec.best);
    if (localexmlraw) {
        this.localeSet(CSL.setupXml(localexmlraw), langspec.best, langspec.best);
    } else {
        if (!localexmlraw && langspec.base !== langspec.best) {
            localexmlraw = this.sys.retrieveLocale(langspec.base);
            if (localexmlraw) {
                this.localeSet(CSL.setupXml(localexmlraw), langspec.base, langspec.best);
            }
        } else if (!localexmlraw && langspec.best !== "en-US") {
            localexmlraw = this.sys.retrieveLocale("en-US");
            if (localexmlraw) {
                this.localeSet(CSL.setupXml(localexmlraw), "en-US", langspec.best);
            }
        }
    }
```

I don't know the implications of making a change like this, as I don't have a good grasp of the full system yet. But I wondered if this was a change you might look positively upon. If so, I'd be happy to issue a pull request.

**Comment by fbennett** [comment]
@NateWr Sorry for the very long delay in getting back to you on this. The patch itself causes lots of breakage across the test suite, but I've made a smaller change to remove the unconditional load of the en-US locale, which may help.

Locale loading is a little squirrelly though, and there are cases where providing only the file corresponding to the declared locale will not work. There is a mapping table of the files associated with each of the "base" (two-character) locales, and when a variant of the "primary" locale is used (such as en-GB), the primary locale associated with the "base" is loaded (which would be en-US), followed by the "best" local (en-GB). Another issue is that in multilingual styles supported by Juris-M, there can be additional locales apart from the default.

To minimize the amount of code that needs moving around, the processor contains a [function for sniffing the locales associated with a style](https://github.com/Juris-M/citeproc-js/blob/master/src/util_locale_sniff.js), the logic of which might be useful for your needs (assuming this is still a live issue in your project).

Closing this -- but if you have any questions, feel free to shout. I'll try to be better with the response time on the next iteration.
